SVG Style Tag

kaalleen opened the pull request:
Elements which define their styles only in the style tag of the svg were ignored by Ink/Stitch and users would see the message, that Ink/Stitch doesn't know how to handle the element.

This can be fixed by using the style attribute getter from inkex. But this won't work for elements which are generated on the fly or taken from other documents (e.g. lettering). An other disadvantage is the loss of the cached style attributes.

@lexelby do you know a better way to handle this or do you think it is ok as is?

lexelby commented:
Hmm.  I'm going to have to read this more closely on my computer to really understand what's going on here.

I wonder, why do elements generated on the fly fail?  Is it because we're creating lxml.etree.Element objects directly?  Can we fix it by creating elements using the svg element wrapper class that inkex uses?

kaalleen commented:
> Can we fix it by creating elements using the svg element wrapper class that inkex uses?

Uh, I don't know?

kaalleen commented:
> Can we fix it by creating elements using the svg element wrapper class that inkex uses?

@lexelby can you lay hands on it or point me to how this could be done?

lexelby commented:
Sure!  It might be a few days.

lexelby commented:
It looks like we should be creating elements using the classes inkex provides, like `my_path = PathElement()`, and then insert them into their parent with `parent.insert(my_path)`.  That way we'll get all the inkex element stuff like `my_path.cascaded_style()`.  This would let us stop catching AttributeError.

That still leaves the question of how to handle stuff that is detached from an SVG, like in Lettering.  I think the right thing to do is catch FragmentError and use `composed_style()`.

lexelby commented:
Unfortunately it looks like `cascaded_style()` doesn't handle inheriting styles from parent elements, so this PR would still be a step back.  :(  A month ago, the inkex folks added full support for _correct_ inheritance and `<style>` tags in `element.specified_style()` ([inkex merge request 294](https://gitlab.com/inkscape/extensions/-/merge_requests/294)).  I'm thinking we want to use this as soon as it's stable.

Until then, I think we want to write something that calls `.cascaded_style()` on every parent and add the styles up, ending with the element itself.  If we get FragmentError, fall back to `.composed_style()`. 

This is kind of a mess.  I can't wait until inkex does all this for us.

kaalleen commented:
Thank you for working on this issue. I really hope too, that this will not be necessary soon. I am not sure if you finished your work, but when I run the testfile, it is trying to render path855 with a fill, which it shouldn't.

lexelby commented:
Oh interesting, I missed that.  I wonder why it's doing that...

kaalleen commented on lib/elements/element.py:
```diff
@@ -194,8 +194,40 @@ def _get_style_raw(self, style_name):
+
+        def _cascaded_composed_style(node):
+            parent = node.getparent()
+            if parent is not None and isinstance(parent, inkex.ShapeElement):
```
Are groups shape elements?

lexelby replied:
I think so.

lexelby replied:
I stole this code directly from inkex ;)

kaalleen commented:
It's not parsing the inline styles (although cascaded_style claims it does).
I updated the branch (kinda ugly code) to parse at least stroke and fill inline styles.
I am not sure how to handle it better. Maybe it's worth waiting until inkex has been updated?

kaalleen commented:
This is working nicely with specified_style now (as far as I can tell). Thank you for the update.
The lettering system though is still using the fallback to the former raw style - as long as auto-route satin is involved.

lexelby commented:
I think I know how to deal with that.  We just need to make sure that XML element objects are always in an SvgDocumentElement.  We can just create one to store each glyph in.  I think that should avoid the need for the fallback entirely.

lexelby commented:
Not quite what I had in mind... I was thinking we could actually store the glyph in an SvgDocumentElement, in the Glyph object.  Is there a reason I'm missing that that would not work?

lexelby commented:
I _think_ we may not even need the SvgDocumentElement workaround.  I just pushed a version that uses inkex element classes instead of plain etree elements, and it seems to work well.  Do you see any cases where we get FragmentError?

Looks like we have conflicts.  I'll take a look at those.  It also seems like somehow I've broken PDF Export, so we need to figure out what's wrong before we can merge...

lexelby commented:
Okay, conflicts fixed.

kaalleen commented:
> It also seems like somehow I've broken PDF Export, so we need to figure out what's wrong before we can merge...

It did not render anything at all ;) Since the color block groups where created but not inserted into the svg.

I like your new method much better. I didn't want to use the svg document in the lettering because I really didn't like the extra layer (the extra svg document within the document). We could've removed it afterwards, but that would have been a little bit too much for what we've tried to achieve. Since we can now use the inkex elements directly, that is really nice.

lexelby commented:
> It did not render anything at all ;) Since the color block groups where created but not inserted into the svg.

I don't understand... What do you mean?

lexelby commented:
Oh, I get it now, thank you!  I must have forgotten that one when I switched away from etree.SubElement().

kaalleen commented:
Do you think this is ready to be merged?

lexelby commented:
I'm not sure... I didn't test it much.  Have you been able to verify whether I fixed the issues we were having with lettering and auto-route satin?

I also haven't tested a release built from this branch to make sure that the new inkex version works properly with pyinstaller.  Have you tried it?

kaalleen commented:
It fixed the auto-route satin within the lettering. Haven't tested the release though. Will do.

kaalleen commented:
Bad and good news. Linux and windows is working. MacOS is not. But the macOS build seems to be broken, but this goes for all recent branches, not just this one. Also locally build macOS "release versions" are working.

lexelby commented:
Darn.  Time to debug builds again.

kaalleen commented:
Yes. It's the pyinstaller version. I will post an update tomorrow or the day after tomorrow.

kaalleen commented:
Ok, I rebased this branch because of merge issues within the lettering extension (we changed the ensure_layer method).

Then I downgraded pyinstaller for macOS to version 4.3. The newest version has a feature which we really may like to use in the future, but for now it is breaking our build. In future it probably will make signing a bit easier for us.

dbalabka commented:
@kaalleen , @lexelby I'm getting the following error in my case. It seems that attribute `specified_style` might not exists in some cases
```
  File "/Users/torinaki/www/htdocs/embroidery/lib/extensions/print_pdf.py", line 300, in effect
    if not self.get_elements():
  File "/Users/torinaki/www/htdocs/embroidery/lib/extensions/base.py", line 176, in get_elements
    self.elements = nodes_to_elements(self.get_nodes(troubleshoot))
  File "/Users/torinaki/www/htdocs/embroidery/lib/extensions/base.py", line 173, in get_nodes
    return self.descendants(self.document.getroot(), troubleshoot=troubleshoot)
  File "/Users/torinaki/www/htdocs/embroidery/lib/extensions/base.py", line 159, in descendants
    nodes.extend(self.descendants(child, selected, troubleshoot))
  File "/Users/torinaki/www/htdocs/embroidery/lib/extensions/base.py", line 141, in descendants
    if (node.tag in EMBROIDERABLE_TAGS or node.tag == SVG_GROUP_TAG) and element.get_style('display', 'inline') is None:
  File "/Users/torinaki/www/htdocs/embroidery/lib/elements/element.py", line 144, in get_style
    style = self._get_specified_style().get(style_name, default)
  File "/Users/torinaki/www/htdocs/embroidery/lib/elements/element.py", line 141, in _get_specified_style
    return self.node.specified_style()
  File "/Users/torinaki/www/htdocs/embroidery/.venv/lib/python3.7/site-packages/inkex/elements/_base.py", line 151, in __getattr__
    raise AttributeError("Can't find attribute {}.{}".format(self.typename, name))
AttributeError: Can't find attribute PathElement.specified_style
```

lexelby commented:
Wow, I honestly have no idea how that's possible.  specified_style is an instance method that's defined in the same class as the __getattr__ that's emitting that exception.  Is something _deleting_ the specified_style attribute from this PathElement instance?

@dbalabka Could you please open a new issue for this?  If you can share an SVG that triggers this, that would be really awesome.

dbalabka commented:
@lexelby thanks for the quick reply. Probably, it is related to the way I'm using plugins code. I'm calling it directly in my CLI command. So it is might be related to the version of  `inkex` or how I load the SVG file. So, I assume that the problem is on my end. Still, with version `v2.0.0` everything works as expected. I will investigate this first.

lexelby commented:
Oh, it's almost certainly due to the inkex version.  We need a quite recent one to get the specified_style method.  Check requirements.txt to see which version we need.

lexelby commented:
I'm really curious what you're up to, btw :)
